Implement Media3

tzugen opened the pull request:
ToDO:

- [x] implement Reordering/removing playlist items
- [x] migrate Auto calls
- [x] Fix Widget 
- [x] Need new shuffle icon
- [x] Use uptodate api endpoints (Known issue: You need to restart the app after changing server)
- [x] NowPlayinbg button does nothing

Normal playback etc. should(!) work.

tzugen committed:
Don't manage ID manually, but use autoGenerate to ensure uniqueness
(across the lifetime of the db)

tzugen committed:
Switch to Media3

tzugen committed:
Remove unused file

tzugen committed:
Remove gapless setting

tzugen committed:
Split Cache into dedicated datasource

tzugen committed:
Detekt fixes

tzugen committed:
Fix a bug in MetadataParser (unrelated to media3)

tzugen committed:
Enable HW playback for better performance

tzugen committed:
Migrate AutoMediaBrowser

tzugen committed:
Fix insertAfterCurrent, Fix getUri

tzugen committed:
Remove a bunch of now unused prefs

tzugen committed:
Add pref switch for HW offload

tzugen committed:
Use grayed out shuffle and repeat buttons.
Fix removing and resorting playlist.

tzugen committed:
Formating / detekt

tzugen committed:
AppWidget to Kotlin

tzugen committed:
Fix Widget

tzugen committed:
Format

tzugen committed:
Remove unused resources

tzugen committed:
Remove unused resources & update lint baseline

hmueller01 committed:
fixed copy paste error with repeat button

ogarcia commented on gradlew:
```diff
@@ -1,7 +1,7 @@
 
 #
-# Copyright � 2015-2021 the original authors.
+# Copyright � 2015-2021 the original authors.
```
Seems that there is a codification problem with this file

tzugen replied:
Iiek! Fixed

tzugen committed:
Fix memory settings

tzugen committed:
Make full body logging an optional debug setting.

tzugen committed:
Remove GC logging

tzugen committed:
Solve more simple...

tzugen committed:
Set memory to 3.2g

tzugen committed:
Rework ActiveServer handling.
Remove blocking call on setting the server.
Implement offline server display more cleanly.
Reconfigure the SourceFactory when the active server has changed

tzugen committed:
Fix a small bug in togglePlayPause()

tzugen committed:
Move alpha value to const val.

hmueller01 reviewed (commented):
Tx.. That's better. Was thinking about that as well.

hmueller01 reviewed (approved):
Yes, this fixes the widget play button problem I was facing ...

tzugen committed:
Merge remote-tracking branch 'origin/develop' into media3-flat

tzugen committed:
Add some more logging calls

tzugen committed:
Fix a bunch of problems with the DataSources

tzugen committed:
Minor

tzugen committed:
Cleanup

tzugen committed:
Squashed commit of the following:

commit c7d9af4529bd9bfba49c46208819d7ef53542cd4
Author: Holger Müller <github@euhm.de>
Date:   Sat Apr 16 17:56:41 2022 +0200

    fixed CI failure

commit 59e18bc5c0264a4a41ead2d819142458c8053ee8
Author: Holger Müller <github@euhm.de>
Date:   Sat Apr 16 17:42:49 2022 +0200

    removed unused dataSourceFactory, commented debug output

tzugen committed:
Squashed commit of the following:

commit 7a8f20ade7d2522c1281b63a31c3e9630df12553
Author: Holger Müller <github@euhm.de>
Date:   Fri Apr 8 23:21:15 2022 +0200

    make build working with target sdk 31

commit 5890c32b7eb91621ce628ac6d900cb00a1481d71
Author: Holger Müller <github@euhm.de>
Date:   Fri Apr 8 22:57:23 2022 +0200

    make build working with target sdk 31

commit 1c7c4839b3eed5eef8298c2ddce22812e14d6329
Merge: 48b0cdea 92ef78a3
Author: Holger Müller <github@euhm.de>
Date:   Fri Apr 8 21:55:14 2022 +0200

    Merge remote-tracking branch 'upstream/media3-flat' into media3-flat

commit 48b0cdea83433cc29760972c522a497f812d6d30
Merge: 9101980c 3ca25ed1
Author: Holger Müller <github@euhm.de>
Date:   Fri Apr 8 19:58:55 2022 +0200

    Merge remote-tracking branch 'upstream/media3-flat' into media3-flat

commit 9101980cb6d8972ee6e49b58737b36c4ac36e495
Merge: 515690ab 97798446
Author: Holger Müller <github@euhm.de>
Date:   Fri Apr 8 17:07:54 2022 +0200

    Merge remote-tracking branch 'upstream/media3-flat' into media3-flat

commit 515690abaccd5405709b1fe877a9ef9c9bbf3e21
Author: Holger Müller <github@euhm.de>
Date:   Thu Apr 7 08:55:42 2022 +0200

    made button off mode better visible

tzugen committed:
Remove legacyPlayerState

tzugen committed:
Cleanup restore

tzugen committed:
Enable StrictMode logging, rm unused prefs

tzugen committed:
Cleanup MediaPlayerLifecycleSupport.kt

tzugen committed:
launch restore on Main Thread

tzugen committed:
Fix de(serilization)

tzugen committed:
Use previous values for channel id

tzugen committed:
Cleanup

tzugen committed:
Fix prefs

tzugen committed:
Always call prepare on adding items

tzugen committed:
- Set Wake mode flag according to Offline status
- Use Rx to trigger CheckDownloads()
- Fix #680 by listening to PositionDiscontinuity
- Throttle RxEvents

tzugen committed:
Remove throttle again, causes calls from wrong thread...

tzugen committed:
Ensure main thread usage in Rx subscription.

tzugen committed:
RM unused PlayerState.kt

tzugen committed:
Create lock before launching the coroutine,
return always a copy of the list and not the real one.

nitehu commented on ultrasonic/src/main/kotlin/org/moire/ultrasonic/service/PlaybackStateSerializer.kt:
```diff
@@ -43,8 +43,8 @@ class PlaybackStateSerializer : KoinComponent {
 
-        ioScope.launch {
-            if (lock.tryLock()) {
+        if (lock.tryLock()) {
```
This won't work, the unlock will be called from another thread which throws an exception.

tzugen committed:
Try to fix the mess :)

nitehu commented on ultrasonic/src/main/kotlin/org/moire/ultrasonic/playback/LegacyPlaylistManager.kt:
```diff
@@ -0,0 +1,108 @@
+    @JvmField
+    var currentPlaying: DownloadFile? = null
+
+    private val mediaItemCache = LRUCache<String, DownloadFile>(1000)
```
So this is where the size of the playlist is limited to 1000.
I've just run into a dead end, and as I see from the comments, you've already been there.
I've tried to remove this cache, and add the DownloadFile directly into the Tag property of the MediaItem.
Unfortunately somehow the LocalConfiguration get stripped from the MediaItems, and the Tag is lost...
I'd really wouldn't like to limit the playlist to 1000. Do you have any better ideas, how this could be handled properly?

tzugen replied:
OK that makes sense now.

 I think to solve it real properly would be to remove this class altogether and rework the Recyclerviews to deal with MediaItems. But that would be too large to put it into this PR.

The easiest fix would probably be to increase the capacity. I just put 1000 because I didn't expect anyone to have that large Playlist. Or try to recreate the DownloadFile based on the values from the MediaItem. But writing this converter would be wasted work later when the class is Refactor Ed away

In the meantime

nitehu replied:
Ok that's what I was thinking too.
I'll increase it to 2000 and continue testing.
(I love to use Ultrasonic as my private radio station with all my favorite songs loaded into it, that's ~1500 songs now, but I also think there are few ppl out there who would load such a list... :)

nitehu committed:
Incremented max playlist size
Fixed DownloadService start
Minor fixes

nitehu commented on ultrasonic/src/main/kotlin/org/moire/ultrasonic/service/Downloader.kt:
```diff
@@ -1,89 +1,107 @@
+                Timber.e(all, "checkDownloads() failed.")
+            } finally {
+                if (!shouldStop) {
+                    Handler(Looper.getMainLooper()).postDelayed(this, CHECK_INTERVAL)
```
For me it seems every time the downloadChecker is started, it will create a loop of itself which checks every five seconds. In a short time there will be tens of downloadChecker instances which will all check in every five seconds, flooding everything.
I think if checking in every five seconds is necessary, there should be two checkers: one that is run every five seconds, and one that can be started ad-hoc, and will run only once.

tzugen replied:
This should not happen!
We should keep a reference to the runnable and if it's active don't create a new one....

nitehu committed:
Fixed Notification Ids to be different for Downloader and Player
Fixed multiple start of periodic CheckDownloads

nitehu committed:
Fixed Lint errors

tzugen committed:
Readd Headphone plug and Bluetooth listeners

tzugen committed:
Add exported attributes, fix lint

nitehu committed:
Fixed not closed connection in CachedDataSource

nitehu commented on ultrasonic/src/main/kotlin/org/moire/ultrasonic/playback/CachedDataSource.kt:
```diff
@@ -27,7 +27,7 @@ import timber.log.Timber
 class CachedDataSource(
     private var upstreamDataSource: DataSource
-) : BaseDataSource(false) {
+) : BaseDataSource(true) {
```
We should check what this does in the Media3 code, but there is a possibility for this to go to the network so I think it is better to signal that

nitehu commented on ultrasonic/src/main/kotlin/org/moire/ultrasonic/playback/CachedDataSource.kt:
```diff
@@ -186,6 +186,8 @@ class CachedDataSource(
             responseByteStream?.close()
             responseByteStream = null
+        } else {
+            upstreamDataSource.close()
```
But the real culprit was that the SSL connection wasn't closed in the underlying ApiDataSorce.
I haven't tested this fully but now the freezing and the timeouts seem to be gone.

tzugen replied:
Wow so proud you found it 😊!

nitehu committed:
Fixed AndroidManifest to receive Bluetooth events
Set parallel downloads to 2 so the third thread can stream for playback

tzugen committed:
Pause playback when swiping away...

tzugen committed:
Merge remote-tracking branch 'origin/develop' into media3-flat

tzugen committed:
Always call prepare before starting playback,
otherwise resuming playback is difficult when the session was dismissed...

tzugen committed:
Update copyright

tzugen committed:
Shutdown service on exit button press

ogarcia reviewed (changes requested):
The app generally works well on both Android 9 and 10 (which are the devices I have), but, right now, it's an app you can't close.

If you dismiss the app, the music stops and the notification remains, that is correct behavior, but if you dismiss the notification, it appears again. You need to do it twice for the notification not to appear again, and even then the app is still running in the background.

If you press exit in the menu, the app closes, but does not exit, it remains in the background.

From what I see, the way other apps solve these problems is that the moment you dismiss the app, the notification closes as well. That is, there can't be a notification if the app is not running in the background.

What do you think?

tzugen committed:
Don't ecit the process, and reinit Koin on resume

tzugen committed:
Call stopForeground(yes)

tzugen committed:
Fix reappearing notification

tzugen committed:
Completely stop and release the player on dismissing the app.

tzugen committed:
Clear widget when service is being shutdown

tzugen committed:
Reset widget on manual exit as well.
